### PR TITLE
fix(base-driver): Update/simplify the logic for logger prefix

### PR DESCRIPTION
**Pull request by mykola-mokhnach**
## Proposed changes

This restores https://github.com/appium/appium/pull/16636 with some slight changes to make ts linter happy

### CODE REVIEW

**Comment by boneskull** [comment]
Can you explain what this avoids?

**Comment by mykola-mokhnach** [comment]
This avoid possible memory leak in case the logger object itself is preserved somewhere. Without WeakRef the actual driver instance would be preserved with it as it is captured in the corresponding closure.

File: packages/base-driver/lib/basedriver/core.js [hunk 1]
```diff
@@ -153,14 +153,26 @@ class DriverCore {
...
+        if (self instanceof DriverCore) {
+          sessionId = self.sessionId;
+        } else if (global.WeakRef && self instanceof global.WeakRef) {
+          const ref = self.deref();
+          if (ref?.sessionId) {
+            sessionId = ref.sessionId;
+          }
+        }
+        return sessionId ? `${instanceName} (${sessionId.substring(0, 8)})` : instanceName;
+      });
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/core.js:175]: I don't think `WeakRef` is necessary here.  In fact, we should probably [avoid use](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef#avoid_where_possible) unless there is no other possible way to do so.

What we're trying to do here is hand a prefix to the logger.  Currently, we're passing a function, but it can just be a string.  This prefix may be dynamic, based upon the current `sessionId` (if any).

What if, upon session creation/destruction (in `createSession` and `deleteSession` of `BaseDriverCore`), we did something like this:

```js
// this would also probably need to be in AppiumDriver's createSession(), which is *not the same* function
async createSession(stuff) {
  this.sessionId = getNewSessionId();
  const logPrefix = createLogPrefix() // does the instanceName stuff w/ the ternary operator, etc.

  // we could do this:
  this._log = logger.getLogger(logPrefix);

  // OR refactor the logger to expose a `setPrefix` method:
  this._log.setPrefix(logPrefix);
}
```

We could create a setter for `sessionId` which did the log-prefix dance as well.

This would only create a potential leak if a driver implementation held on to the old value `this._log` for too long.  But a driver could create a potential leak in myriad ways which we cannot prevent.

**Review by boneskull — CHANGES_REQUESTED** [review]
I don't think we need `WeakRef` here... please see comment.

File: packages/base-driver/lib/basedriver/core.js (hunk 1 above)
Comment (boneskull) [reply packages/base-driver/lib/basedriver/core.js:175]: If we do this, we should also add a test that shows the prefix gets updated.

Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/core.js]: changed to add/remove the prefix in create/deleteSession

**Comment by mykola-mokhnach** [comment]
I have updated the implementation, but now I have no idea about how to update unit tests after _log object gets reassigned. I also don't have a good idea on how to make prefix mutable on the logger itself

**Comment by boneskull** [comment]
@mykola-mokhnach Can I push to your branch?

**Comment by mykola-mokhnach** [comment]
> @mykola-mokhnach Can I push to your branch?

sure, not a problem at all

**Comment by mykola-mokhnach** [comment]
Seems like I got some progress as of now. Please review once more @boneskull

**Review by jlipps — APPROVED** [review]

File: packages/types/src/index.ts [hunk 2]
```diff
@@ -365,6 +365,7 @@ export interface DriverHelpers {
   isPackageOrBundle: (app: string) => boolean;
   duplicateKeys: <T>(input: T, firstKey: string, secondKey: string) => T;
   parseCapsArray: (cap: string | string[]) => string[];
+  generateDriverLogPrefix: (obj: Object, sessionId?: string) => string;
```
Comment (boneskull) [inline packages/types/src/index.ts:368]: This is weird, and I had to look it up.

`Object` allows primitives, but not `null` nor `undefined`.  While that may actually work, it's not the intent of this function.  You probably mean `object`, which means an "actual" object.  

And again, while that might actually work, the only thing we really expect `obj` to be is an object implementing `Core`.  So I think `Object` should be `Core`.

I'm also not sure why some things go in "driver helpers" and not others.  But that's neither here nor there.

> However--and I view this as a major shortcoming of TypeScript--neither `object` nor `Object` understand if the object in question extends `Object.prototype`.  This means, given `const obj: object = Object.create(null)` _or_ `const obj: Object = Object.create(null)`  and then `generateDriverLogPrefix(obj)`, TS will _not_ complain that the `constructor` property does not exist on `obj` (it doesn't), and you will get a runtime error.

File: packages/base-driver/lib/basedriver/helpers.js [hunk 3]
```diff
@@ -546,6 +546,22 @@ function parseCapsArray (cap) {
...
+ */
+function generateDriverLogPrefix (obj, sessionId = null) {
+  const instanceName = `${obj.constructor.name}@${node.getObjectId(obj).substring(0, 4)}`;
+  return sessionId ? `${instanceName} (${sessionId.substring(0, 8)})` : instanceName;
+}
+
 /** @type {import('@appium/types').DriverHelpers} */
-export default {configureApp, isPackageOrBundle, duplicateKeys, parseCapsArray};
-export {configureApp, isPackageOrBundle, duplicateKeys, parseCapsArray};
+export default {
```
Comment (boneskull) [inline packages/base-driver/lib/basedriver/helpers.js:562]: We should probably move these "double" exports across the codebase; they are not needed anywhere except maybe on the public API edges.  (just saying)

File: packages/types/src/index.ts [hunk 4]
```diff
@@ -481,10 +482,15 @@ export interface EventHistoryCommand {
 }
 export type HTTPMethod = _Method;
 
+export type PrefixCallback = () => string;
+
+export type Prefix = PrefixCallback|string;
```
Comment (boneskull) [inline packages/types/src/index.ts:487]: If we _do_ keep `PrefixCallback`, I think it's fine to shorten this to just

```suggestion
export type Prefix = string | () => string;
```

and

**Review by boneskull — COMMENTED** [review]
This is good and was basically what I was suggesting, so thanks.  

For a test I think all you really need to do is have something that compares `this.log.prefix` before and after calling `createSession()`, and again for `deleteSession()`.  We can assert the prefix looks a certain way, or we can just assert it's not the same.

You'd probably want to put that in the `baseDriverUnitTests` function of `test/basedriver/driver-tests.js`

File: packages/types/src/index.ts (hunk 2 above)
Comment (mykola-mokhnach) [reply packages/types/src/index.ts:368]: changed type to Core

File: packages/base-driver/lib/basedriver/helpers.js (hunk 3 above)
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/helpers.js:562]: it was like that before, so it does not make sense to me to change it in this PR

File: packages/types/src/index.ts (hunk 4 above)
Comment (mykola-mokhnach) [reply packages/types/src/index.ts:487]: removed prefixcallback

File: packages/base-driver/lib/basedriver/core.js [hunk 5]
```diff
@@ -102,9 +102,9 @@ class DriverCore {
   eventEmitter = new EventEmitter();
 
   /**
-  * @type {AppiumLogger?}
+  * @type {AppiumLogger}
   */
-  _log = null;
+  _log = logger.getLogger(helpers.generateDriverLogPrefix(this));
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/base-driver/lib/basedriver/core.js:107]: makes sense. Moved the initializer to the class constructor

File: packages/types/src/index.ts [hunk 6]
```diff
@@ -481,10 +482,15 @@ export interface EventHistoryCommand {
...
+
 export interface AppiumLogger {
   unwrap(): Logger;
   level: string;
   levels: string[];
+  prefix?: Prefix,
```
[the remark this answers is not in the document]
Comment (mykola-mokhnach) [reply packages/types/src/index.ts:493]: it was like that before. Not sure I need to change that now

File: packages/base-driver/test/basedriver/driver-tests.js [hunk 7]
```diff
@@ -55,6 +55,12 @@ function baseDriverUnitTests (DriverClass, defaultCaps = {}) {
       sessId.length.should.be.above(5);
     });
 
+    it('should setup log prefix after createSession', async function () {
+      const previousPrefix = d.log.prefix;
+      await d.createSession(null, null, w3cCaps);
+      previousPrefix.should.not.eql(d.log.prefix);
+    });
```
Comment (boneskull) [inline packages/base-driver/test/basedriver/driver-tests.js:62]: Don't we need to reset the prefix after `deleteSession` is called?  Are we doing that?  If not, we need to do that, I think. And add a test.

It'd be more valuable to make an assertion about the expected value (since there's logic involved in determining the prefix) instead of just asserting it has changed (in the above test case and also one for `deleteSession`)

Comment (mykola-mokhnach) [reply packages/base-driver/test/basedriver/driver-tests.js]: updated the test
